### PR TITLE
Add Swagger Documentation for PdfGenerator2122Controller

### PR DESCRIPTION
**Pull request by opticbob**
## Summary

- *This work is behind a feature toggle (flipper): NO*
- *(Summarize the changes that have been made to the platform):* The Swagger documentation for `RepresentationManagement::V0::PdfGenerator2122Controller` has been added in the `rswag` format and the existing documentation for the `RepresentationManagement::V0::PowerOfAttorneyController` has been converted to that same format.
- *(If bug, how to reproduce)*
- *(What is the solution, why is this the solution?)*
- *(Which team do you work for, does your team own the maintenance of this component?):* FAR/ARM
- *(If introducing a flipper, what is the success criteria being targeted?)*

## Related issue(s)

- [Add Swagger Documentation for RepresentationManagement::V0::PdfGenerator2122Controller#91500](https://github.com/department-of-veterans-affairs/va.gov-team/issues/91500)
- *Link to previous change of the code/bug (if applicable)*
- *Link to epic if not included in ticket*

## Testing done

- [ ] *New code is covered by unit tests:* No, this is a documentation only change.
- *Describe what the old behavior was prior to the change*
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing:* With the application up and running you can load `http://localhost:3000/representation_management/v0/apidocs` in a web browser and see the docs in JSON format.  You can also follow the instructions [here](https://github.com/department-of-veterans-affairs/vets-api/tree/master/app/swagger) to load the swagger viewer then point it at that same url to get a more interactive view.
- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots
<details>
<summary>Tall screenshot of entire `RepresentationManagement `apidocs page</summary>

Screenshot 2024-10-08 at 15-13-33 Swagger UI

</details>

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas):* It only touches the `RepresentationManagement` apidocs page.

## Acceptance criteria

- [x]  I added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [x]  I added a screenshot of the developed feature

### CODE REVIEW

**Comment by opticbob** [comment]
<details>
<summary>JSON output of the apidocs page</summary>
```
{
  "swagger": "2.0",
  "info": {
    "version": "1.0.0",
    "title": "va.gov Representation Management API",
    "description": "The API for managing representation for VA Forms 21-22 and 21-22a",
    "termsOfService": "",
    "contact": {
      "name": "va.gov team"
    },
    "license": {
      "name": "Creative Commons Zero v1.0 Universal"
    }
  },
  "tags": [
    {
      "name": "PDF Generation",
      "description": "Generate a PDF form from user input"
    }
  ],
  "host": "127.0.0.1:3000",
  "schemes": [
    "https",
    "http"
  ],
  "basePath": "/",
  "consumes": [
    "application/json"
  ],
  "produces": [
    "application/json"
  ],
  "parameters": {
    "authorization": {
      "name": "Authorization",
      "in": "header",
      "description": "The authorization method and token value",
      "required": true,
      "type": "string"
    },
    "optional_authorization": {
      "name": "Authorization",
      "in": "header",
      "description": "The authorization method and token value",
      "required": false,
      "type": "string"
    }
  },
  "paths": {
    "/representation_management/v0/pdf_generator2122": {
      "post": {
        "summary": "Generate a PDF for form 21-22",
        "operationId": "createPdfForm2122",
        "produces": [
          "application/pdf"
        ],
        "tags": [
          "PDF Generation"
        ],
        "parameters": [
          {
            "name": "pdf_generator2122",
            "in": "body",
            "description": "Form data for generating PDF",
            "required": true,
            "schema": {
              "type": "object",
              "properties": {
                "organization_name": {
                  "type": "string",
                  "example": "Veterans Organization"
                },
                "record_consent": {
                  "type": "boolean",
                  "example": true
                },
                "consent_address_change": {
                  "type": "boolean",
                  "example": false
                },
                "consent_limits": {
                  "type": "array",
                  "items": {
                    "type": "string"
                  },
                  "example": [
                    "ALCOHOLISM",
                    "DRUG_ABUSE",
                    "HIV",
                    "SICKLE_CELL"
                  ]
                },
                "conditions_of_appointment": {
                  "type": "array",
                  "items": {
                    "type": "string"
                  },
                  "example": [
                    "a123",
                    "b456",
                    "c789"
                  ]
                },
                "claimant": {
                  "type": "object",
                  "properties": {
                    "name": {
                      "type": "object",
                      "properties": {
                        "first": {
                          "type": "string",
                          "example": "John"
                        },
                        "middle": {
                          "type": "string",
                          "example": "A"
                        },
                        "last": {
                          "type": "string",
                          "example": "Doe"
                        }
                      }
                    },
                    "address": {
                      "type": "object",
                      "properties": {
                        "address_line1": {
                          "type": "string",
                          "example": "123 Main St"
                        },
                        "address_line2": {
                          "type": "string",
                          "example": "Apt 1"
                        },
                        "city": {
                          "type": "string",
                          "example": "Springfield"
                        },
                        "state_code": {
                          "type": "string",
                          "example": "IL"
                        },
                        "country": {
                          "type": "string",
                          "example": "US"
                        },
                        "zip_code": {
                          "type": "string",
                          "example": "62704"
                        },
                        "zip_code_suffix": {
                          "type": "string",
                          "example": "1234"
                        }
                      }
                    },
                    "date_of_birth": {
                      "type": "string",
                      "format": "date",
                      "example": "12/31/2000"
                    },
                    "relationship": {
                      "type": "string",
                      "example": "Spouse"
                    },
                    "phone": {
                      "type": "string",
                      "example": "1234567890"
                    },
                    "email": {
                      "type": "string",
                      "example": "veteran@example.com"
                    }
                  }
                },
                "veteran": {
                  "type": "object",
                  "properties": {
                    "insurance_numbers": {
                      "type": "array",
                      "items": {
                        "type": "string"
                      },
                      "example": [
                        "123456789",
                        "987654321"
                      ]
                    },
                    "name": {
                      "type": "object",
                      "properties": {
                        "first": {
                          "type": "string",
                          "example": "John"
                        },
                        "middle": {
                          "type": "string",
                          "example": "A"
                        },
                        "last": {
                          "type": "string",
                          "example": "Doe"
                        }
                      }
                    },
                    "address": {
                      "type": "object",
                      "properties": {
                        "address_line1": {
                          "type": "string",
                          "example": "123 Main St"
                        },
                        "address_line2": {
                          "type": "string",
                          "example": "Apt 1"
                        },
                        "city": {
                          "type": "string",
                          "example": "Springfield"
                        },
                        "state_code": {
                          "type": "string",
                          "example": "IL"
                        },
                        "country": {
                          "type": "string",
                          "example": "US"
                        },
                        "zip_code": {
                          "type": "string",
                          "example": "62704"
                        },
                        "zip_code_suffix": {
                          "type": "string",
                          "example": "1234"
                        }
                      }
                    },
                    "ssn": {
                      "type": "string",
                      "example": "123456789"
                    },
                    "va_file_number": {
                      "type": "string",
                      "example": "123456789"
                    },
                    "date_of_birth": {
                      "type": "string",
                      "format": "date",
                      "example": "12/31/2000"
                    },
                    "service_number": {
                      "type": "string",
                      "example": "123456789"
                    },
                    "service_branch": {
                      "type": "string",
                      "example": "Army"
                    },
                    "service_branch_other": {
                      "type": "string",
                      "example": "Other Branch"
                    },
                    "phone": {
                      "type": "string",
                      "example": "1234567890"
                    },
                    "email": {
                      "type": "string",
                      "example": "veteran@example.com"
                    }
                  }
                }
              }
            }
          }
        ],
        "responses": {
          "200": {
            "description": "PDF generated successfully"
          },
          "422": {
            "description": "unprocessable entity response",
            "schema": {
              "$ref": "#/definitions/Errors"
            }
          }
        }
      }
    },
    "/representation_management/v0/power_of_attorney": {
      "get": {
        "summary": "Get Power of Attorney",
        "description": "Retrieves the Power of Attorney for a veteran, if any.",
        "operationId": "getPowerOfAttorney",
        "tags": [
          "Power of Attorney"
        ],
        "responses": {
          "200": {
            "description": "Successfully checked for Power of Attorney information. Returns Power of Attorney details if available; otherwise, returns an empty object.",
            "schema": {
              "$ref": "#/definitions/PowerOfAttorneyResponse"
            }
          },
          "404": {
            "description": "Resource not found",
            "schema": {
              "type": "object",
              "properties": {
                "errors": {
                  "type": "array",
                  "items": {
                    "type": "object",
                    "properties": {
                      "title": {
                        "type": "string",
                        "example": "Resource not found"
                      },
                      "detail": {
                        "type": "string",
                        "example": "Resource not found"
                      },
                      "code": {
                        "type": "string",
                        "example": "404"
                      },
                      "status": {
                        "type": "string",
                        "example": "404"
                      }
                    }
                  }
                }
              }
            }
          },
          "500": {
            "description": "Unexpected server error",
            "schema": {
              "type": "object",
              "properties": {
                "errors": {
                  "type": "object",
                  "properties": {
                    "title": {
                      "type": "string",
                      "example": "Internal server error"
                    },
                    "detail": {
                      "type": "string",
                      "example": "Unexpected error occurred"
                    },
                    "code": {
                      "type": "string",
                      "example": "500"
                    },
                    "status": {
                      "type": "string",
                      "example": "500"
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  "definitions": {
    "PowerOfAttorneyResponse": {
      "type": "object",
      "properties": {
        "data": {
          "type": "object",
          "properties": {
            "id": {
              "type": "string",
              "example": "123456"
            },
            "type": {
              "type": "string",
              "description": "          Specifies the category of Power of Attorney (POA) representation. This field differentiates between two primary forms of POA:\n          - 'veteran_service_representatives': Represents individual representatives who are authorized to act on behalf of veterans. These representatives include attorneys and claim agents.\n          - 'veteran_service_organizations': Denotes organizations accredited to provide representation to veterans.\n",
              "enum": [
                "veteran_service_representatives",
                "veteran_service_organizations"
              ]
            },
            "attributes": {
              "required": [
                "type",
                "name",
                "address_line1",
                "city",
                "state_code",
                "zip_code"
              ],
              "type": "object",
              "properties": {
                "type": {
                  "type": "string",
                  "description": "Type of Power of Attorney representation",
                  "example": "organization",
                  "enum": [
                    "organization",
                    "representative"
                  ]
                },
                "name": {
                  "type": "string",
                  "example": "Veterans Association"
                },
                "address_line1": {
                  "type": "string",
                  "example": "1234 Freedom Blvd"
                },
                "city": {
                  "type": "string",
                  "example": "Arlington"
                },
                "state_code": {
                  "type": "string",
                  "example": "VA"
                },
                "zip_code": {
                  "type": "string",
                  "example": "22204"
                },
                "phone": {
                  "type": "string",
                  "example": "555-1234"
                },
                "email": {
                  "type": "string",
                  "example": "contact@example.org"
                }
              }
            }
          }
        }
      }
    },
    "Errors": {
      "required": [
        "errors"
      ],
      "properties": {
        "errors": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/Error"
          }
        }
      }
    },
    "Error": {
      "required": [
        "title",
        "detail",
        "code",
        "status"
      ],
      "properties": {
        "title": {
          "type": "string",
          "example": "Bad Request",
          "description": "error class name"
        },
        "detail": {
          "type": "string",
          "example": "Received a bad request response from the upstream server",
          "description": "possibly some additional info, or just the class name again"
        },
        "code": {
          "type": "string",
          "example": "422",
          "description": "Sometiems just the http code again, sometimes something like \"EVSS400\", where\" \\\n                    \" the code can be found in config/locales/exceptions.en.yml"
        },
        "source": {
          "type": [
            "string",
            "object"
          ],
          "example": "RepresentationManagement::V0::PdfConstructor::Form2122",
          "description": "error source class"
        },
        "status": {
          "type": "string",
          "example": "422",
          "description": "http status code"
        }
      }
    }
  }
}
```
</details>

**Comment by ryan-mcneil** [comment]
Please merge master, it should fix the failing spec 🤞

File: modules/representation_management/app/swagger/errors.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    end
+  end
+
+  swagger_schema :Error do
+    key :required, %i[title detail code status]
+    property :title, type: :string, example: 'Bad Request', description: 'error class name'
+    property :detail, type: :string, example: 'Received a bad request response from the upstream server',
+                      description: 'possibly some additional info, or just the class name again'
+    property :code, type: :string, example: '422',
+                    description: 'Sometiems just the http code again, sometimes something like "EVSS400", where" \
```
Comment (jvcAdHoc) [inline modules/representation_management/app/swagger/errors.rb:23]:
```suggestion
                    description: 'Sometimes just the http code again, sometimes something like "EVSS400", where" \
```

File: modules/representation_management/config/routes.rb [hunk 2]
```diff
@@ -11,5 +11,6 @@
     resources :pdf_generator2122a, only: %i[create]
     resources :power_of_attorney, only: %i[index]
     get 'apidocs', to: 'apidocs#index'
+    resources :apidocs2, only: [:index]
```
Comment (jvcAdHoc) [inline modules/representation_management/config/routes.rb:14]: Left over from testing?

Comment (opticbob) [reply modules/representation_management/config/routes.rb]: Fixed!

File: modules/representation_management/spec/support/rswag_config.rb [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                    }
+                  }
+                }
+              }
+            }
+          }
+        },
+        paths: {},
+        # basePath is not valid OAS v3
+        servers: [
```
Comment (jvcAdHoc) [inline modules/representation_management/spec/support/rswag_config.rb:125]: Probably want staging in here too

File: modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb
Comment (jvcAdHoc) [inline modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb:1]: 422 is a possible response

File: modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                'code' => 'abc'
+              }
+            }
+          }
+          allow_any_instance_of(BenefitsClaims::Service).to receive(:get_power_of_attorney).and_return(lh_response)
+        end
+
+        schema anyOf: [
+          {
+            '$ref' => '#/components/schemas/PowerOfAttorneyResponse'
```
Comment (jvcAdHoc) [inline modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb:38]: This schema doesn't match the full shape that `modules/representation_management/app/serializers/representation_management/power_of_attorney/base_serializer.rb` allows for

File: modules/representation_management/spec/support/rswag_config.rb [hunk 5]
```diff
@@ -0,0 +1,158 @@
+# frozen_string_literal: true
+
+class RepresentationManagement::RswagConfig
+  def config # rubocop:disable Metrics/MethodLength
+    {
+      'modules/representation_management/app/swagger/v0/swagger.json' => {
+        openapi: '3.0.1',
+        info: {
+          title: 'va.gov Representation Management API',
+          version: '1.0.0',
```
Comment (jvcAdHoc) [inline modules/representation_management/spec/support/rswag_config.rb:10]: Probably want to use `v0` and `0.1.0` where applicable to be consistent with our namespace

File: modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb
Comment (opticbob) [reply modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb:1]: Good catch, I added it.

File: modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb (hunk 4 above)
Comment (opticbob) [reply modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb:38]: I added the missing attributes.  You can see them below:
Screenshot 2024-10-07 100709

File: modules/representation_management/spec/support/rswag_config.rb (hunk 3 above)
Comment (opticbob) [reply modules/representation_management/spec/support/rswag_config.rb:156]: Yep, it's been added.

File: modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb (hunk 4 above)
Comment (jvcAdHoc) [reply modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb:39]: This is very nit-picky, but the top level type of `veteran_service_representatives` isn't possible with the sub type `organization`

File: modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        schema '$ref' => '#/components/responses/InternalServerError'
+
+        run_test!
+      end
+
+      response '500', 'Internal Server Error' do
+        schema '$ref' => '#/components/responses/InternalServerError'
+        run_test!
+      end
```
Comment (jvcAdHoc) [inline modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb:66]: something about the config isn't showing the full error response object:
image

Just getting the `"string"`

Also, the 422 response shouldn't be an internal server error

Comment (opticbob) [reply modules/representation_management/spec/requests/representation_management/v0/power_of_attorney_swagger_spec.rb]: The full error response is now working correctly:
Screenshot 2024-10-08 151501

You can see the whole thing by expanding the screenshot accordion in the PR description.  It's up to date.

File: modules/representation_management/spec/support/rswag_config.rb [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+class RepresentationManagement::RswagConfig
+  def config # rubocop:disable Metrics/MethodLength
+    {
+      'modules/representation_management/app/swagger/v0/swagger.json' => {
+        openapi: '3.0.1',
+        info: {
+          title: 'va.gov Representation Management API',
+          version: '0.1.0',
+          termsOfService: 'https://developer.va.gov/terms-of-service',
+          description: 'The API for managing representation for VA Forms 21-22 and 21-22a'
```
Comment (holdenhinkle) [inline modules/representation_management/spec/support/rswag_config.rb:12]: I think the description should reflect the broader range of functionalities provided by these APIs. Something like, 'A set of APIs powering the POA Widget, Find a Representative, and Appoint a Representative.' @jvcAdHoc What do you think?

File: rakelib/rswag.rake [hunk 8]
```diff
@@ -63,6 +63,16 @@ namespace :rswag do
       run_tasks_in_parallel(%w[rswag:appeals_api:prod rswag:appeals_api:dev])
     end
   end
+
+  namespace :representation_management do
+    desc 'Generate rswag docs for the representation_management'
```
Comment (holdenhinkle) [inline rakelib/rswag.rake:68]: Perhaps change this to `'Generate rswag docs for representation_management'` (i.e. remove the word 'the').

**Review by holdenhinkle — COMMENTED** [review]
Looks good! Just a couple of minor comments.

File: rakelib/rswag.rake (hunk 8 above)
Comment (opticbob) [reply rakelib/rswag.rake:68]: Good catch!

File: modules/representation_management/spec/support/rswag_config.rb (hunk 7 above)
Comment (opticbob) [reply modules/representation_management/spec/support/rswag_config.rb:12]: Sounds good to me!

Comment (opticbob) [reply modules/representation_management/spec/support/rswag_config.rb]: Screenshot 2024-10-08 154209

**Review by holdenhinkle — DISMISSED** [review]
LGTM!

**Review by stevenjcumming — CHANGES_REQUESTED** [review]
Can you resolve the `RSpec/ScatteredSetup` and `Metrics/MethodLength` cop warnings?

**Comment by opticbob** [comment]
@stevenjcumming I resolved all the `Metrics/MethodLength` messages but was unable to resolve the `RSpec/ScatteredSetup` message.  Any way I tried to unify the setup in one `before do` block would ultimately break.  I am matching the implementation present in the `appeals_api` and `claims_api` rswag specs anywhere they have to mock/stub something.

Do you have any ideas for how I can move forward?

**Review by holdenhinkle — APPROVED** [review]

**Comment by stevenjcumming** [comment]
> @stevenjcumming I resolved all the `Metrics/MethodLength` messages but was unable to resolve the `RSpec/ScatteredSetup` message. Any way I tried to unify the setup in one `before do` block would ultimately break. I am matching the implementation present in the `appeals_api` and `claims_api` rswag specs anywhere they have to mock/stub something.
> 
> Do you have any ideas for how I can move forward?

I'll have to look at it later, but I can look passed it for this PR.

**Review by stevenjcumming — APPROVED** [review]

**Pull request merged**